### DISCUSSION TITLE
Question (module): How does one access module options?

### DISCUSSION THREAD

**Discussion by mohsin** [nuxt/nuxt#16341] [Questions]
Question related to `Branch: module`

So I've defined [the module options](https://github.com/nuxt/starter/blob/module/src/module.ts) like this in the config:
```
export default defineNuxtModule({
  meta: {
    name: 'my-module',
    configKey: 'mymodule'
  },
  defaults: {
    foo: 'bar'
  },
  hooks: {
    'components:dirs'(dirs) {
      dirs.push({
        path: fileURLToPath(new URL('./components', import.meta.url))
      })
    }
  }
})
```

And in my project, I'm using it like:
```
export default defineNuxtConfig({
    modules: [
      'my-module',
    ],
    mymodule: {
      foo: 'someothervalue'
    },
})
```

As you can see in the original module setup, I'm using components import for custom components. Now I want to use the config value of `foo` inside my component's `<script setup lang="ts">` so I can take that variable and pass it to a the component template's element attribute. How do I go about doing this?

**Comment by dsvgl** [comment]
As you are in your nuxt.config, you could declare it once and give it to the module to make some decisions by it. And also eg use runtimeConfig to have the same value globally available.

Simplified example:
```js
// nuxt.config.ts
const MY_FOO = 'someothervalue';

export default defineNuxtConfig({
  modules: ['my-module'],
  runtimeConfig: {
    public: {
      foo: MY_FOO,
    },
  },
  mymodule: {
    foo: MY_FOO,
  },
});
```

```vue
<template>
  <div>foo is {{ foo }}</div>
</template>

<script setup lang="ts">
const { foo } = useRuntimeConfig().public;
</script>
```

**Reply by mohsin** [reply]
Thanks for the response. This is the same solution I thought of but it’s not too good because:

1. Redundancy—The same value is being used in public config and module config.
2. I could just skip using the module config entirely and just check runtime config on my module’s component (fallback to default otherwise).

I was hoping for a more standard method / solution for this if it exists.

**Comment by dsvgl** [comment]
Ah yeah, your second suggestions is also good. That's actually how I use it one of my projects.
I think you could also create/add a nuxt-plugin from within your module that exposes your var by using `provide` inside the template. Actually adding the var to your plugin dynamically can be done (I think) via [`addPluginTemplate`](https://v3.nuxtjs.org/api/advanced/kit#plugins). Not really good documentation for this, but maybe you can look into the code of other modules. I saw it being used.
But yeah, no standard method as far as I know.

**Reply by mohsin** [reply]
> But yeah, no standard method as far as I know.

That's disappointing. I hoped there was an undocumented standardized method to get the value. I actually did find out one other workaround while experiementing to get it besides the solution you described earlier, and it goes like:
```
<script setup>
...
import { loadNuxtConfig } from '@nuxt/kit'
const foo = (await loadNuxtConfig()).mymodule.foo;
...
```

So when I put this in my component.vue, it does give me the value i.e. `someothervalue` from the nuxt config. However, I get a warning ` ERROR  This module cannot be imported in the Vue part of your app. [importing @nuxt/kit from ../dist/components/MyComponent.vue]` as well because I don't think `@nuxt/kit` was designed to be used inside Vue components. I really hope that a simplified solution is found.

Anyway, thanks a lot for your inputs. I'll keep this discussion open so I can hopefully find a more elegant solution to this problem.

**Comment by IsraelOrtuno** [comment]
I wasn't able to make this work but Nuxt Content Module seems to basically write the properties right to the `runtimeConfig` from the module itself:

https://github.com/nuxt/content/blob/main/src/module.ts#L568

I have tried this but for some reason they get overwritten when rendering. Probably missing one more step or something. I hope this helps.

**Reply by dsvgl** [reply]
Have you also used defu or someting similar to merge existing config with new values. Good approach btw.

**Reply by IsraelOrtuno** [reply]
I haven't, was actually testing it quickly to find a solution for this discussion. If using `defu` solves it would be good to know :)

**Comment by Vinccool96** [comment]
I had enough [and decided to create a feature request](https://github.com/nuxt/framework/issues/9192)
